pgx/src/stringinfo.rs

```diff
@@ -189,6 +191,48 @@ impl StringInfo {
         }
     }
 
+    /// Reads a range of bytes, modifying the underlying cursor to reflect what was read
+    ///
+    /// Returns None if the underlying remaining binary is smaller than requested with the range.
+    ///
+    /// Ranges can start from an offset, resulting in skipped information.
+    ///
+    /// Most common use-case for this is reading the underlying data in full (`read(..)`)
+    pub fn read<R: RangeBounds<usize>>(&mut self, range: R) -> Option<&[u8]> {
+        use std::ffi::c_int;
+        let remaining = unsafe { (*self.sid).len - (*self.sid).cursor } as usize;
+        let start = match range.start_bound() {
+            Bound::Included(bound) => *bound,
+            Bound::Excluded(bound) => *bound + 1,
+            Bound::Unbounded => 0,
+        };
+        let end = match range.end_bound() {
+            Bound::Included(bound) => *bound,
+            Bound::Excluded(bound) => *bound - 1,
+            Bound::Unbounded => remaining,
+        };
+        let total = end - start;
+
+        if total > remaining {
+            return None;
+        }
+
+        // safe:  self.sid will never be null
+        Some(unsafe {
+            if (*self.sid).data.is_null() {
+                &[]
+            } else {
+                (*self.sid).cursor += start as c_int;
+                let result = std::slice::from_raw_parts(
+                    (*self.sid).data.add((*self.sid).cursor as usize) as *const u8,
+                    total,
+                );
+                (*self.sid).cursor += total as c_int;
+                result
+            }
+        })
+    }
```

Comment:
We should probably just `impl std::io::Read` using similar underlying code, which can be extracted into another PR. `str` implements a similar [`.get(..)`](https://doc.rust-lang.org/std/primitive.str.html#method.get) but it doesn't modify cursors. I wouldn't mind seeing this function with a different name.

Reply:
I can totally do this, but I think it would make most sense to do this after #903